File: src/dependencies/crazyflie_hardware/src/crtp_cpp/src/crtp_cpp/packer/link_layer_packer.cpp

#include "crtp_cpp/packer/link_layer_packer.hpp"
#include <cstring>

#define PORT_LINK 0xF

#define CHANNEL_ECHO 0
#define CHANNEL_SOURCE 1
#define CHANNEL_SINK 2
#define CHANNEL_NULL 3

#define CHANNEL_LINK 0xF

#define BOOTLOADER_HEADER 0xFE
#define BOOTLOADER_CMD_ALLOFF 0x01
#define BOOTLOADER_CMD_SYSOFF 0x02
#define BOOTLOADER_CMD_SYSON 0x03
#define BOOTLOADER_CMD_GETVBAT 0x04
#define BOOTLOADER_CMD_RESET_INIT 0xFF
#define BOOTLOADER_CMD_RESET 0xF0

#define BOOTLOADER_RESET_TO_BOOTLOADER 0
#define BOOTLOADER_RESET_TO_FIRMWARE 1

LinkLayerPacker::LinkLayerPacker()
    : CrtpPacker(PORT_LINK) {}

CrtpPacket LinkLayerPacker::prepare_packet(uint8_t channel, const std::vector<uint8_t>& data) {
    return CrtpPacker::prepare_packet(channel, data);
}

CrtpPacket LinkLayerPacker::echopacket() {
    std::vector<uint8_t> data;
    return prepare_packet(CHANNEL_ECHO, data);
}

CrtpPacket LinkLayerPacker::sourcepacket() {
    std::vector<uint8_t> data;
    return prepare_packet(CHANNEL_SOURCE, data);
}

CrtpPacket LinkLayerPacker::sinkpacket() {
    std::vector<uint8_t> data;
    return prepare_packet(CHANNEL_SINK, data);
}

CrtpPacket LinkLayerPacker::nullpacket() {
    std::vector<uint8_t> data;
    return prepare_packet(CHANNEL_NULL, data);
}

CrtpPacket LinkLayerPacker::get_vbat() {
    std::vector<uint8_t> data = {BOOTLOADER_HEADER, BOOTLOADER_CMD_GETVBAT};
    return prepare_packet(CHANNEL_LINK, data);
}

CrtpPacket LinkLayerPacker::platform_power_down() {
    std::vector<uint8_t> data = {BOOTLOADER_HEADER, BOOTLOADER_CMD_ALLOFF};
    return prepare_packet(CHANNEL_LINK, data);
}

CrtpPacket LinkLayerPacker::stm_power_down() {
    std::vector<uint8_t> data = {BOOTLOADER_HEADER, BOOTLOADER_CMD_SYSOFF};
    return prepare_packet(CHANNEL_LINK, data);
}

CrtpPacket LinkLayerPacker::stm_power_up() {
    std::vector<uint8_t> data = {BOOTLOADER_HEADER, BOOTLOADER_CMD_SYSON};
    return prepare_packet(CHANNEL_LINK, data);
}

CrtpPacket LinkLayerPacker::reset_init() {
    std::vector<uint8_t> data = {BOOTLOADER_HEADER, BOOTLOADER_CMD_RESET_INIT};
    return prepare_packet(CHANNEL_LINK, data);
}

CrtpPacket LinkLayerPacker::reset_to_bootloader() {
    std::vector<uint8_t> data = {BOOTLOADER_HEADER, BOOTLOADER_CMD_RESET, BOOTLOADER_RESET_TO_BOOTLOADER};
    return prepare_packet(CHANNEL_LINK, data);
}

CrtpPacket LinkLayerPacker::reset_to_firmware() {
    std::vector<uint8_t> data = {BOOTLOADER_HEADER, BOOTLOADER_CMD_RESET, BOOTLOADER_RESET_TO_FIRMWARE};
    return prepare_packet(CHANNEL_LINK, data);
}

File: src/dependencies/crazyflie_hardware/src/crtp/crtp/packers/link_layer_packer.py
from .packer import Packer
from .crtp_packer import CrtpPacker

import struct

from typing import Callable


class LinkLayerPacker(Packer):
    PORT_LINK = 0xF

    CHANNEL_ECHO = 0
    CHANNEL_SOURCE = 1
    CHANNEL_SINK = 2
    CHANNEL_NULL = 3

    CHANNEL_LINK = 0xF

    BOOTLOADER_HEADER = 0xFE
    BOOTLOADER_CMD_ALLOFF = 0x01
    BOOTLOADER_CMD_SYSOFF = 0x02
    BOOTLOADER_CMD_SYSON = 0x03
    BOOTLOADER_CMD_GETVBAT = 0x04
    BOOTLOADER_CMD_RESET_INIT = 0xFF
    BOOTLOADER_CMD_RESET = 0xF0

    BOOTLOADER_RESET_TO_BOOTLOADER = 0
    BOOTLOADER_RESET_TO_FIRMWARE = 1

    def __init__(self, crtp_packer_factory: Callable[[int], CrtpPacker]):
        super().__init__(crtp_packer_factory, self.PORT_LINK)

    def echopacket(self):
        data = struct.pack("<")
        return self._prepare_packet(channel=self.CHANNEL_ECHO, data=data)

    def sourcepacket(self):
        data = struct.pack("<")
        return self._prepare_packet(channel=self.CHANNEL_SOURCE, data=data)

    def sinkpacket(self):
        data = struct.pack("<")
        return self._prepare_packet(channel=self.CHANNEL_SINK, data=data)

    def nullpacket(self):
        data = struct.pack("<")
        return self._prepare_packet(channel=self.CHANNEL_NULL, data=data)

    def get_vbat(self):
        data = struct.pack("<BB", self.BOOTLOADER_HEADER, self.BOOTLOADER_CMD_GETVBAT)
        return self._prepare_packet(channel=self.CHANNEL_LINK, data=data)

    def platform_power_down(self):
        data = struct.pack("<BB", self.BOOTLOADER_HEADER, self.BOOTLOADER_CMD_ALLOFF)
        return self._prepare_packet(channel=self.CHANNEL_LINK, data=data)

    def stm_power_down(self):
        data = struct.pack("<BB", self.BOOTLOADER_HEADER, self.BOOTLOADER_CMD_SYSOFF)
        return self._prepare_packet(channel=self.CHANNEL_LINK, data=data)

    def stm_power_up(self):
        data = struct.pack("<BB", self.BOOTLOADER_HEADER, self.BOOTLOADER_CMD_SYSON)
        return self._prepare_packet(channel=self.CHANNEL_LINK, data=data)

    def reset_init(self):
        data = struct.pack(
            "<BB", self.BOOTLOADER_HEADER, self.BOOTLOADER_CMD_RESET_INIT
        )
        return self._prepare_packet(channel=self.CHANNEL_LINK, data=data)

    def reset_to_bootloader(self):
        """
        Resets the cf to bootloader, reset_init has to be called before this
        """
        data = struct.pack(
            "<BBB",
            self.BOOTLOADER_HEADER,
            self.BOOTLOADER_CMD_RESET,
            self.BOOTLOADER_RESET_TO_BOOTLOADER,
        )
        return self._prepare_packet(channel=self.CHANNEL_LINK, data=data)

    def reset_to_firmware(self):
        """
        Resets the cf to firmware, reset_init has to be called before this
        """
        data = struct.pack(
            "<BBB",
            self.BOOTLOADER_HEADER,
            self.BOOTLOADER_CMD_RESET,
            self.BOOTLOADER_RESET_TO_FIRMWARE,
        )
        return self._prepare_packet(channel=self.CHANNEL_LINK, data=data)
